docs/proposals/custom-sidecar-config.md

```diff
@@ -0,0 +1,200 @@
+# Custom Container Configuration
+
+## Introduction
+Users and field team have requested access to custom configuration of kubernetes sidecar and init containers.
+See [#3925](https://github.com/kumahq/kuma/issues/3925).
+This intuitively aligns with the ProxyTemplate configuration,
+which provides a free-form configuration of Envoy.
+The current plan is to introduce a new CRD,
+similar to ProxyTemplate,
+which can accommodate sidecar config.
+
+## Custom Resource Definition
+The new CRD will be named ContainerTemplate.
+It will allow for customer configuration of
+both sidecar and init containers.
+
+This will follow a similar format to ProxyTemplate.
+A ContainerTemplate will match pods via namespaces and labels,
+and modify the containers associated with those pods.
+The spec will contain full templates for “sidecar” and “init” dataplane containers.
+
+### Example
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: ContainerTemplate
+metadata:
+  name: container-template-1
+spec:
+  selectors:
+    - namespace: namespace-1
+  sidecarTemplate: | …
+  initTemplate: | …
+```
+
+## Dynamic Configuration
+Kuma creates the existing container specifications, in part,
+using dynamic data known at container creation time.
+Since the full configuration will be specified in the CRD,
+we need a method for these dynamic elements to be provided by kuma,
+but also potentially overridden by the user.
+We will achieve this by providing the user with a full container template
+where the dynamic values are specified as template variables.
+If the user needs to override the value,
+they can replace the variable with their own literal override.
+Otherwise, it will be filled in by kuma.
+
+## Error Modes and Validation
+We will validate that the rendered template meets the kubernetes specification.
+We will not validate that it is a sane configuration.
+It is assumed that anyone using this feature has the expertise to create and debug a container configuration
+via interaction with kubernetes directly.
+
+If multiple ContainerTemplate policies match the same container,
+the most recently applied policy wins.
+
+## Templates
+Template specification will be the full container specification,
+with variables for Kuma to fill with default values.
+Ideally, the user can modify individual values,
+and upgrades which affect unrelated values will just fill in those variables
+without user intervention.
+Larger changes across versions will require user to integrate their changes into a new template.
+
+### Example - Init Container Default
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: ContainerTemplate
+metadata:
+  name: container-template-1
+spec:
+  selectors:
+    - namespace: namespace-1
+  initTemplate: |
+    name: kuma-init
+    command: {{ .Command }}
```

Comment:
How does a user know all these options?

Reply:
The idea is that we give them the default template, which has all the available vars pre-filled for them to change. 

Now, as I was actually building this, it became evident that pretty much the entire template needs to be variables, which in most cases is just a more complicated way to express "change this one element, but leave the rest to kuma". When we discussed the original design (which was more like proxy template - specify elements to update) with the broader team + field, they were big on "just give us a place to write the full description". I think I would like someone from the field to chime in on what we settled on now that it is fully rendered (@johnharris85 ?)

Reply:
I've added back the original template design to the end for comparison.